### ISSUE TITLE
maven provider plugin read classpathElements from main class instead of test 

### ISSUE THREAD

**Issue by cosineyan** [DiUS/pact-jvm#285]
When doing the pact provider verification for message type, it seems maven plugin always scan code from src/main, which such kind of code usually are testing code. 

Within pact-jvm\pact-jvm-provider-maven\src\main\groovy\au\com\dius\pact\provider\maven\PactProviderMojo.groovy, classpathElements  parameter is defined by default as ${project.runtimeClasspathElements} with readonly=true.

Suggest to allow this parameter for configuration within client's pom.xml.

**Comment by uglyog** [comment]
I have defaulted it to the test classpath and also allowed it to be overridden.

**Comment by cosineyan** [comment]
A further question: 
I saw the concept is to use reflection utility to scan the classes for functions with annotation to be run. 
If this annotated function calls/utilizes any other complex type/proc within the project but out of the class scope, reflection utility will definitely fail with exception of something like "no class definition found". What do you think about it? Does it mean those plugins might not be adequate for using for message pact verification, and JUnit will be a better way (as it's integrated with provider's code and probably have all symbols in hand)?

**Comment by uglyog** [comment]
Not sure what you mean by 'as it's integrated with provider's code', but the plugin is executed in the context of a classpath, much the same as the surefire plugin which runs the tests. By default, the test classpath includes the compile classpath.

The utility used to scan the test classes just finds the annotated methods, it does not need all dependencies resolved. Once found, the methods are then executed with the test classpath from the project, so as long as all dependencies are present it will work.

In testing this, I have found that the test classpath was not resolved correctly, so I needed to add an additional annotation to the Mojo.

**Comment by cosineyan** [comment]
Yeah tomorrow I'll have another try. 
Previously I tried to fix the classpath issue locally in maven plugin and call a verification but failed with NoClassDefFound.
However if I directly put provider as a dependency and call ProviderVerfier myself - the target annotated proc was able to resolve other dependency. 
Maybe I fixed the classpath wrongly? Tomorrow I will fetch your fixing code and see whether the result differs.
